---
title: "Best friends: what collection is friendly to this tag? Is it actually friendly?"
output: 
  html_vignette:
vignette: >
 %\VignetteIndexEntry{Best friends: what collection is friendly to this tag? Is it actually friendly?}
 %\VignetteEngine{knitr::rmarkdown}
 %\VignetteEncoding{UTF-8}
 %\VignetteDepends{devtools}
---
<!--  tried below html_vignette, did no work
      theme: Bootstrap
      toc: true
      toc_depth: 2 -->


[SI AUGUSTUS CERNATUR, CERNANTUR QOUQUE AMICI]{style="float:right"}

`r format(Sys.Date(), "%d %b %Y")` 

`r library(knitr)`
`r library(markdown)`
`r library(best.friends)`
`r options(width=80,keep.source=TRUE)`
`r opts_chunk$set(echo=TRUE,message=FALSE,cache=FALSE,warning=FALSE)`
`r set.seed(42)`
<!-- cache is off, we save all the necessary things in Rda -->

<div style="line-height: 1.5em;">


# Introduction
There is a simple intuition of what it means to be a friend. A friend of Augustus cares about Augustus more than about other people. And, if we see Augustus, then we infer to see friends(s) of Augustus also. Letâ€™s translate it into statistical language.

Consider a set of genes and their loads in a set of expression patterns. Each pattern represents a biological process by the expression levels of the involved genes. 

Sometimes, the expression of a single gene indicates the activity of the entire pattern. In the simplest case, the gene has a nonzero load in only one pattern. 
Moreover, the gene may have several nonzero loads, but all of them but one are relatively small. Then the gene (AKA Augustus) is the marker gene for the pattern, and the marked pattern is the best friend of this gene. We want to identify the marker genes and corresponding patterns statistically.  

\paragraph{Model} The bipartite graph naturally fits this model. To generalise the example, we will refer to genes as \textit{{\tag}s}, and to expression patterns, as \textit{{\collection}s}, and to any load as \textit{attention}. The sketch below illustrates the setting.

![The bipartite graph picture shows only edges with nonzero values](bipartite.png){width=25%}

To be more specific, we are given a set of collections $C = \{c_1, \dots, c_k\}$, and a set of tags $T = \{t_1, \dots, t_n\}$.
Each tag $t \in T$ and collection $c \in C$ are related by the attention $A(t, c)$, that the cloud pays to the tag. The strength of attention varies from $0$ to infinity, $A(t, c)\ge 0$. 

In the following, the attention values are stored in $n\times k$ matrix $\mathcal{A}$. 

The tag-collection-attention model applies to many problems in bioinformatics and statistics.

        Example                           tag $t$         collection $c$                           attention $A\left(t,c\right)$
--------------------------------------   -------------   ----------------------------------  ----------------------------
[gene regulation by TFs](#toyTF)          gene            genes under the TF regulation       strength of regulation
[transctironal correlations](#toynet)     gene            genes coexpressed with a gene       transcription correlation
fuzzy clustering                          object          cluster                             object weight in cluster
transcription decomposition               transcript      transcription pattern               transcript's load in pattern
weighted graph                            vertex          another vertex                      weight of edge between collection and tag
--------------------------------------   -------------   ----------------------------------  ----------------------------



\paragraph{Goal} For each tag, we want to identify the collection(s) that particularly prefer(s) the tag, if any. Then such a cloud is a friend (or the best friend if there is only one) for the tag. The simplest example: imagine that only one cloud pays attention to the tag in hand. We intend to express the meaning of the word "particularly" in this context by a statistical test.

Let's denote the rank of the $i$-th tag's attention value paid by $j$-the collection inside the collection, or, in our notation, rank of $A_{ij}$ in $A_{*j}$ column, as $r(t_i,c_j)$. 

The main idea of the test is that if a tag $t_i$ is now preferred by any collection, the distribution of the ranks of the tag in all the collections $r(t_i,c_j)$ behaves like $j$  i.i.d integers that are uniformely distributed in $1..n$


Please see our manuscript [] for more details.


## Toy example: Bayesian option
```{r source, echo=FALSE}
data.digits=2
p.val.digits=4
```
</div>


Let's illustrate the method with something supersimple.

```{r}
regulation<-structure(c(
  0.2, 0.25, 0.1, 0.23, 0.3, 0.12, 0.14, 
  4, 3, 6, 1, 9, 7, 10, 
  3, 1, 5, 6, 3, 4, 11), 
  dim = c(7, 3), 
  dimnames = list(
    c("Gene1", "Gene2", "Gene3", 
      "Gene4", "Gene5", "Gene6", "Gene7"), 
    c("TF1", "TF2", "TF3")))
noquote(regulation)
```
What we want to identify is Gene2-TF1 pair. Gene2 is not the strongest downstream gene in the TF1 collection (column), it is only the second, but, on the other hand its rank in other TF-based collections is much lower.

First, let's apply the Bayesian version of the method, `best.friends.bic`.

```{r}
best.friends.bic(regulation,prior.to.have.friends = .5)
```

We see the expected Gene2-TF1 pair as well as the Gene1-TF1 pair, that is similar to the expected, Gene1 has rank of 3 in the TF1 collection.

The lower is the prior to have friends for a tag (it is an obligatory parameter for `best.friends.bic`), the less marker-friend pairs we get.

```{r}
best.friends.bic(regulation,prior.to.have.friends = .33)
```

```{r}
best.friends.bic(regulation,prior.to.have.friends = .25)
```

In all the three examples above, the `best.no` parameter was 1 by default. It means that the function filters out all the tags that have more that one friendly collection. If we relax the restriction, we get:

```{r}
best.friends.bic(regulation,prior.to.have.friends = .5, best.no = "all")
best.friends.bic(regulation,prior.to.have.friends = .33, best.no = "all")
best.friends.bic(regulation,prior.to.have.friends = .25, best.no = "all")
```

## The same toy example: default option

Let's run the p-value-based option of the method on the same data. It is less sensitive on small dataset, so we use unpractical threshold for the p-value to see the result.

```{r}
best.friends(regulation,threshold = .5,p.adjust.method = "no")
```

Amd, again, if we relax the `best.no` restriction, we will see more pairs, but now the friends are not required to be unique for a marker.

```{r}
best.friends(regulation,threshold = .5,p.adjust.method = "no",best.no = "all")
```

## Session info
```{r}
sessionInfo()
```
